Tom boi

Friziepop opened the pull request:
Merge me pl0x

Friziepop committed:
added my implementations

Friziepop committed:
refactored stuff

Friziepop committed:
removed unneeded test file

Friziepop committed:
removed unnesseccary stuff

Friziepop committed:
readded in out

Friziepop committed:
fixed a smll bug

Friziepop committed:
fixed warning

amitir22 commented on src/TNode.h:
```diff
@@ -0,0 +1,54 @@
+
+    void setHeight(int height);
+
+private:
```
why private after public? any special reason?

Friziepop replied:
That seemed to be the convention in every case we did projects in c++.
like here for example: https://github.com/Friziepop/mtm-ex3/blob/master/UniqueArray.h

amitir22 replied:
huh... didn't remember then... ok.

amitir22 commented on src/main1.cpp:
```diff
@@ -251,45 +251,45 @@ static errorType OnNumberOfStreams(void* DS, const char* const command) {
-	artists = (int *)malloc(numOfSongs * sizeof(int));
-	songs = (int *)malloc(numOfSongs * sizeof(int));
-
     ValidateRead(sscanf(command, "%d", &numOfSongs), 1, "%s failed.\n", commandStr[GETRECOMMENDEDSONGS_CMD]);
```
did you move it up because of what we talked about? if so just resolve

amitir22 commented on src/main1.cpp:
```diff
@@ -251,45 +251,45 @@ static errorType OnNumberOfStreams(void* DS, const char* const command) {
 
     if (res != SUCCESS) {
         printf("%s: %s\n", commandStr[GETRECOMMENDEDSONGS_CMD], ReturnValToStr(res));
+        if (artists != NULL) free(artists);
+        if (songs != NULL) free(songs);
```
so this is what you were talking about?

Friziepop replied:
Yup

amitir22 commented on src/TNodeImp.h:
```diff
@@ -0,0 +1,80 @@
+#define DATASTRUCTURESWET1_TNODEIMP_H
+
+
+#include "TNode.h"
```
please remove this include ty

Friziepop replied:
Why? Ifdef knows how to deal with cyclic includes

amitir22 replied:
i know.
it will work. 
but why should it be there?

amitir22 commented on src/TNodeImp.h:
```diff
@@ -0,0 +1,80 @@
+using std::make_shared;
+
+template<class Key, class Info>
+TNode<Key, Info>::TNode() = default;
+
+template<class Key, class Info>
+TNode<Key, Info>::~TNode() = default;
```
for cases like those you can write it in the header file 🤔

Friziepop replied:
according to this it's better to implement in the body. As a good practice: https://stackoverflow.com/questions/43733938/default-c-definitions-in-body-or-header-files

amitir22 replied:
hmm... but i think that if we look at the header we want to know if it uses the default ctor aren't we?
if you decide not to change it just resolve

amitir22 commented on src/TNode.h:
```diff
@@ -0,0 +1,54 @@
+
+    void setKey(const Key &newKey);
+
+    int getHeight() const;
```
seems like you wrote getters and then the setters so you might as well move it to the getters part

Friziepop replied:
What?

Friziepop replied:
fiixed

amitir22 commented on src/TNodeImp.h:
```diff
@@ -0,0 +1,80 @@
+}
+
+template<class Key, class Info>
+int TNode<Key, Info>::getHeight() const {
```
if you moved the getter in the header move here too plz

Friziepop replied:
What? There are no getters in the header?

Friziepop replied:
fiiixed

amitir22 commented on src/TNode.h:
```diff
@@ -0,0 +1,54 @@
+
+using std::shared_ptr;
+
+template<class Key, class Info>
```
if we called it key then we should probably rename `Info` to `Value`
so we'll have key-value terminology 😃

Friziepop replied:
fiiixed

amitir22 commented on src/TNodeImp.h:
```diff
@@ -0,0 +1,80 @@
+void TNode<Key, Info>::setInfoPointer(shared_ptr<Info> newInfoPointer) {
+    this->infoPointer = newInfoPointer;
+}
+
```
redundant nl ty

Friziepop replied:
fiiixed

amitir22 commented on src/ArtistInfo.cpp:
```diff
@@ -0,0 +1,45 @@
+namespace DataStructuresWet1 {
+    DataStructuresWet1::ArtistInfo::ArtistInfo(int artistId, int numOfSongs) : artistId(artistId),
+                                                                               numOfSongs(numOfSongs) {
+        this->songs = unique_ptr<SongInfo[]>(new SongInfo[numOfSongs]);
+        this->songsPos = unique_ptr<shared_ptr<BiDirectionalNode<StreamCountGroup>>[]>(
+                new shared_ptr<BiDirectionalNode<StreamCountGroup>>[numOfSongs]);
+        this->songInfoListPointers = unique_ptr<shared_ptr<BiDirectionalNode<SongInfo>>[]>(
+                new shared_ptr<BiDirectionalNode<SongInfo>>[this->numOfSongs]);
```
ok wtf? 😆

amitir22 replied:
hmmm can you extract some variables?

Friziepop replied:
Those variables are extracted... They're just templates so their Type is long

Friziepop replied:
It's kind of problemtic. I'm doing it on purpuse because I want the smart_ptrs to deal with allocation failure automatically. 
If I were to do all of the new allocations before calling the smart_ptr ctors we could have a potential memory leak. 
:(

amitir22 replied:
:( okay...

amitir22 commented on src/SongInfo.cpp:
```diff
@@ -0,0 +1,44 @@
+#include "SongInfo.h"
+
+namespace DataStructuresWet1 {
+    SongInfo::SongInfo() = default;
```
can move to header file

Friziepop replied:
according to this it's better to implement in the body. As a good practice: https://stackoverflow.com/questions/43733938/default-c-definitions-in-body-or-header-files

amitir22 commented on src/SongInfo.cpp:
```diff
@@ -0,0 +1,44 @@
+    SongInfo::SongInfo(int artistId, int songId, int streamCount) : artistId(artistId), songId(songId),
+                                                                    streamCount(streamCount) {}
+
+    SongInfo::~SongInfo() = default;
```
can move to header file

Friziepop replied:
according to this it's better to implement in the body. As a good practice: https://stackoverflow.com/questions/43733938/default-c-definitions-in-body-or-header-files

amitir22 commented on src/BiDirectionalNodeImp.h:
```diff
@@ -0,0 +1,51 @@
+#ifndef DATASTRUCTURESWET1_BIDIRECTIONALNODEIMP_H
+#define DATASTRUCTURESWET1_BIDIRECTIONALNODEIMP_H
+
+#include "BiDirectionalNode.h"
```
please remove

Friziepop replied:
Why not let it sit there? Again that'd what ifndef is for.

amitir22 commented on src/BiDirectionalNodeImp.h:
```diff
@@ -0,0 +1,51 @@
+
+#include "BiDirectionalNode.h"
+
+template<class Type>
+BiDirectionalNode<Type>::BiDirectionalNode() = default;
+
+template<class Type>
+BiDirectionalNode<Type>::~BiDirectionalNode() = default;
```
can be moved to header

Friziepop replied:
According to this: https://stackoverflow.com/questions/43733938/default-c-definitions-in-body-or-header-files
it's better to implement in the body.

amitir22 commented on src/BiDirectionalNode.h:
```diff
@@ -0,0 +1,44 @@
+    Type value;
+    shared_ptr<BiDirectionalNode<Type>> next;
+    // This is of type weak_ptr because otherwise it can create cyclic dependency and create memory leaks if shared_ptr is used
+    weak_ptr<BiDirectionalNode<Type>> previous;
```
amazing idea but it seems like we just hide potential misuse of this class 🤔 
i mean, i want to know if we forgot to reassign a new previous.
we don't want to let previous be `nullptr` because of potential errors.

Friziepop replied:
Nope it's okay. It knows how to deal with nullptr, and that's the use case you use weak ptr for
https://stackoverflow.com/questions/38298008/possible-memory-leaks-with-smart-pointers/51412891

Friziepop replied:
We'd have a memory leak otherwise I checked it myself to make sure there'd be a leak if I didn't use it. 
 it's not a matter of a good idea. It's crucial

amitir22 replied:
if we can code bi-directional linked list in C i'm sure we can code one in cpp with shared pointers. 
it just means we need to be more careful when we implement the list class and how we use the node class.
so i'm not convinced it's CRUCIAL but i'll take it to heart.

amitir22 commented on src/BiDirectionalNodeImp.h:
```diff
@@ -0,0 +1,51 @@
+}
+
+template<class Type>
+void BiDirectionalNode<Type>::setPrevious(shared_ptr<BiDirectionalNode<Type>> newPrevious) {
+    this->previous = newPrevious;
```
are you assigning shared_ptr to weak_ptr?
is that on purpose? does it work?

Friziepop replied:
Yes that's how weak ptr works

Friziepop replied:
I ran a bunch of tests already, it's likely what you see does work

amitir22 commented on src/BiDirectionalNode.h:
```diff
@@ -0,0 +1,44 @@
+
+    Type &getValue();
+
+    shared_ptr<BiDirectionalNode<Type>> getNext() const;
+
+    // Done for optimization when we want to iterate a bit faster
+    BiDirectionalNode<Type> *getNextPointer() const;
```
confusing names and perhaps switched by mistake(?)

amitir22 replied:
if it is switched then we can just 
rename `getNext()` to `getNextPointer2()`
rename `getNextPointer()` to `getNext()`
and remove the 2 from `getNextPointer2()`

amitir22 replied:
ok so after bumping into this a few more times i realized you return a pointer instead of `shared_ptr`.
1. why? is it really that much faster?
2. the getter names are really confusing. 
might want to change `*get<member>Pointer()`
to `*get<member>RawPointer()`
.
why i got confused:
![image](https://user-images.githubusercontent.com/26043145/81995446-1101bd00-9653-11ea-8875-d75159a4bc33.png)

Friziepop replied:
Yes it improved speed by 4 times. In some areas. I did it just in case.

Friziepop replied:
Not that it matter for efficiency, but just incase the segel does some timeout. Most people had their tests run in 3 seconds, this took 12 seconds. Once I switched some places that were executed a lot to use raw pointers we passed it in less than 3 seconds.

Most people used raw pointers so I guess it's a fair comparison.

Friziepop replied:
But I will fix the names

amitir22 replied:
awesome. waiting for the changes

amitir22 replied:
(in all the classes you added it)

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+#define DATASTRUCTURESWET1_BIDIRECTIONALLINKEDLISTIMP_H
+
+#include <cassert>
+#include "BiDirectionalLinkedList.h"
```
ok, please remove all cyclic includes from all the generic ADT's `<ADT>Imp.h` files

Friziepop replied:
But cyclic includes are okay? That's why we have ifdef.

amitir22 replied:
we have ifdef's so when 2 libraries we use share usage of a third library then there will be no conflicts. 🤔 
but if you believe it's more readable then fine

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+#ifndef DATASTRUCTURESWET1_BIDIRECTIONALLINKEDLISTIMP_H
+#define DATASTRUCTURESWET1_BIDIRECTIONALLINKEDLISTIMP_H
+
+#include <cassert>
```
i don't really care if we use asserts here but this is cpp we have exceptions 🤔

Friziepop replied:
Asserts are for programming uses and debugging and are gone once you compile. Without debug flag. Exceptions are differnet

Friziepop replied:
Exceptions are used for users who use your stuff or deal with runtime errors that you can't control.

amitir22 commented on src/BiDirectionalLinkedList.h:
```diff
@@ -0,0 +1,50 @@
+
+    shared_ptr<BiDirectionalNode<Type>> getLast() const;
+
+    void insert(const Type &newValue);
```
i think `append()` is more readable 🤔 
and accurately describes what the methods does

Friziepop replied:
fiiixed

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+        shared_ptr<BiDirectionalNode<Type>> newLast = make_shared<BiDirectionalNode<Type>>();
+
+        newLast->setNext(nullptr);
+        newLast->setPrevious(last);
```
`newLast->setPrevious(last);` 
to 
`newLast->setPrevious(this->last);`
please

amitir22 replied:
or, alternatively, we can remove the `this->`'s from the `if` section.

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+        newLast->setPrevious(last);
+        newLast->setValue(newValue);
+
+        last->setNext(newLast);
```
same as previous comment

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+        last->setNext(newLast);
+
+        this->last = newLast;
+    }
+}
```
i think this version is less duplication:
![image](https://user-images.githubusercontent.com/26043145/81996479-d188a000-9655-11ea-9c56-e0b5a5c9603e.png)

what do you think?

Friziepop replied:
Seems good. Can you copy paste the code? why print screen :(

amitir22 replied:
```
void BiDirectionalLinkedList<Type>::insert(const Type &newValue) {
    shared_ptr<BiDirectionalNode<Type>> newLast = make_shared<BiDirectionalNode<Type>>();
    
    newLast->setValue(newValue);
    newLast->setNext(nullptr);
    
    if (this->first == nullptr) {
        newLast->setPrevious(nullptr);
        
        this->first = newLast;
        this->last = this->first;
    } else {
        newLast->setPrevious(this->last);
        this->last->setNext(newLast);

        this->last = newLast;
    }
}
```

Friziepop replied:
fixxed

amitir22 commented on src/BiDirectionalLinkedList.h:
```diff
@@ -0,0 +1,50 @@
+
+    void insert(const Type &newValue);
+
+    shared_ptr<BiDirectionalNode<Type>> findNodeByValue(const Type &valueToFind) const;
```
we want to allow access to our NODES publicly? 🤨

Friziepop replied:
Yes sometimes why not. It can be useful to get the specific pointer so you can remove it more efficiently

Friziepop replied:
This function is not used anyway I just did it because it's part of the list interface.

Friziepop replied:
Kinda like I purposefully made functions that get a pointer to a node instead of a preallocated value. In order to be able to allocate all memory first then pass allocated smart_ptrs. To deal with potential memory leaks in case of an allocation failure.

amitir22 commented on src/BiDirectionalLinkedList.h:
```diff
@@ -0,0 +1,50 @@
+
+    shared_ptr<BiDirectionalNode<Type>> findNodeByValue(const Type &valueToFind) const;
+
+    void deleteNodeByValue(const Type &valueToFind);
```
if we use the term `remove()` then we should use it here too instead of `delete`

Friziepop replied:
fixxxxxxxxxed

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+    assert(nodeToDelete != nullptr);
+
+    if (nodeToDelete->getNext() != nullptr) {
+        nodeToDelete->getNext()->setPrevious(nodeToDelete->getPrevious().lock());
```
if `nodeToDelete` is the first node then `nodeToDelete->getPrevious().lock()` will be an error wouldn't it?
i think we should move the `if (nodeToDelete == getFirst())` up and put the rest in the else section or something of that effect

Friziepop replied:
There won't be an error I already checked.

Friziepop replied:
Again nullptrs are fine for smart ptrs.

amitir22 replied:
you meant smart pointers are literally EMBEDDED in the cpp language?

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+    if (nodeToDelete == getLast()) {
+        this->last = nodeToDelete->getPrevious().lock();
+    }
+}
```
ok i don't think i understand what happens in this function so maybe my comments are redundant.
i'd like to talk to you about this function 😟

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+        if (valueToFind == current->getValue()) {
+            remove(current);
+
+            break;
```
i would've just used `return;` instead 🤨

Friziepop replied:
fixxxxxxxxxed

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::printList(ostream &os) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
```
why not run from `this->first` to `this->last` and then change
`os << std::endl;`
to
`os << this->last.get() << std::endl;`
i think it's more readable and less "dangerous"

amitir22 replied:
same for `printListReverse()`

amitir22 replied:
ok now isnstead of "dangerous" the only reason for this comment is Readability.
so change it only if you want to. otherwise resolve.

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+void BiDirectionalLinkedList<Type>::printListReverse(ostream &os) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->last; current != nullptr;
+         current = current->getPrevious().lock()) {
+        os << current->getValue() << " ";
```
what will `getValue()` return if `current->getPrevious()` is `nullptr` or expired?

Friziepop replied:
It's okay. Smart ptrs lock know how to deal with nulls
This use case of smart ptrs is alright because we own the entire list and know to put null in a smart ptr. And not let it point to an expired pointer.

Friziepop replied:
You are right though that smart_ptrs can potentially point to expired pointers, however because we control every node we know it won't happen in THIS case.

Friziepop replied:
Of course if an outside user has access to the node pointers he can screw things up, but they shouldn't modify them in the first case. Only pass them around from function to function for efficiency sake.

amitir22 replied:
boi i read about `weak_ptr.lock()`.
what i'm saying is, what if `current->previous` is null? then you'll try to access the method `.get()` from a null pointer.
am i wrong? `nullptr.get()` is a thing?

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+shared_ptr<BiDirectionalNode<Type>>
+BiDirectionalLinkedList<Type>::insertNodeAfter(shared_ptr<BiDirectionalNode<Type>> nodeToInsertAfter,
+                                               shared_ptr<BiDirectionalNode<Type>> newNode) {
+    // If list is empty we insert it to the the head of the list
+    if (getFirst() == nullptr) {
+        assert(nodeToInsertAfter == nullptr);
+
+        this->first = newNode;
+        this->last = newNode;
+
+        return getFirst();
+    }
```
why not use method overloading without `nodeToInsertAfter`?

amitir22 replied:
nvm

amitir22 commented on src/BiDirectionalLinkedListImp.h:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    return newNode;
+}
```
ok this entire function seems suspicious 🤨

Friziepop replied:
Well you need to point out something because this function is necessary for efficiency.

Friziepop replied:
All it does is insert in o(1) inside the list. It's not that complex

amitir22 commented on src/Makefile:
```diff
@@ -0,0 +1,70 @@
+ AVLTreeImp.h BiDirectionalNode.h BiDirectionalNodeImp.h
+	${CC} ${DEBUG_FLAG} ${COMP_FLAG} -c $*.cpp
+	
+MusicManagerTests.o: MusicManagerTests.cpp MusicManager.h AVLTree.h \
```
where are those files? are they git-ignored?

Friziepop replied:
fixxxxxxxxxed

Friziepop committed:
cr

Friziepop committed:
reverted

Friziepop committed:
using -g for debugging

amitir22 commented on src/AVLTree.h:
```diff
@@ -0,0 +1,135 @@
+    shared_ptr<TNode<Key, Info>> leftRotate(shared_ptr<TNode<Key, Info>> treeNode);
+
+    /**
+  * @brief Performs right left rotation on treeNode
```
i think we should mention here which sub-tree we rotate is it the left sub-tree of treeNode->right? 
or the other way around?

amitir22 commented on src/AVLTree.h:
```diff
@@ -0,0 +1,135 @@
+#define DATASTRUCTURESWET1_AVLTREE_H
+
+#include "TNode.h"
+
```
hmm plz look for any sequential new-lines around the project and remove 1 of them pl0x

Friziepop replied:
fiiixxxed

amitir22 commented on src/AVLTree.h:
```diff
@@ -0,0 +1,135 @@
+  *
+  * @return New root after the rotation is done
+  */
+    shared_ptr<TNode<Key, Info>> rightRotate(shared_ptr<TNode<Key, Info>> treeNode);
```
plz change:
`rightRotate()` to `rotateRightChildRight()`
`leftRotate()` to `rotateLeftChildLeft()`
`rightLeftRotate()` to `rotateRightChildLeft()`
`leftRightRotate()` to `rotateLeftChildRight()`

Friziepop replied:
fiiixxxxed

amitir22 replied:
where?

amitir22 commented on src/AVLTree.h:
```diff
@@ -0,0 +1,135 @@
+  * @return Current root after inserting and balancing tree
+  */
+    shared_ptr<TNode<Key, Info>>
+    insertAVLTree(shared_ptr<TNode<Key, Info>> newNode, shared_ptr<TNode<Key, Info>> treeNode);
```
can `treeNode` be renamed to `rootNode`/`subTreeNode`? 
or did i misinterpret the function?

amitir22 replied:
same with `insertAVLTree()` and 
`insertTNode()` 
or 
`insertAVLSubTree()`

amitir22 replied:
(BTW: amazing function)

Friziepop replied:
fixxxed

amitir22 reviewed (commented):
wip (clicked 'start review' by accident)

amitir22 commented on src/AVLTreeImp.h:
```diff
@@ -0,0 +1,345 @@
+
+    updateHeight(treeNode);
+
+    return fixTreeBalance(treeNode);
```
would've added a comment saying "it will be called from leaf to root" or something like that

Friziepop replied:
fixxxxxxxxxed

amitir22 replied:
where?

amitir22 commented on src/AVLTree.h:
```diff
@@ -0,0 +1,135 @@
+using std::shared_ptr;
+
+template<class Key, class Info>
+class AVLTree {
```
is this class not in the same namespace as the rest?
if the problem is the <class>Impl.h file then you can open the same namespace scope there and implement inside it.
i think... 🤔

Friziepop replied:
It's a general code file that we might want to use later in other excercises. According to what I heard. I tried to put it in a namespace and it worked, but clion for some reason screwed the formatting whenever I did it. So I decided to not put it in any namespace like we did in the Mtm exercises in UniqureArray.

amitir22 committed:
small changes:
 - added complexity description for most classes.
 - made some minor readability changes.
 - improved re-useability :D

amitir22 committed:
small changes:
 - updated more complexity functions for private methods in AVLTree.h

amitir22 committed:
cosmetic fixes

amitir22 committed:
cosmetic fixes

Friziepop committed:
finished todos

Friziepop committed:
fixed name

Friziepop committed:
reformatted

Friziepop committed:
Merge pull request #2 from amitir22/amit-cr-tom

WIP: small changes:

Friziepop committed:
cr

Friziepop committed:
cr

Friziepop committed:
cr

Friziepop committed:
cr

Friziepop committed:
cr

amitir22 committed:
cosmetic fixes
